gadget/filesystemimage.go

```diff
@@ -0,0 +1,134 @@
+// positioned structures.
+type FilesystemImageWriter struct {
+	contentDir string
+	ps         *PositionedStructure
+	workDir    string
+}
+
+// PostStageFunc is called after the filesystem contents for the given structure
+// have been staged at a temporary location
+type PostStageFunc func(rootDir string, ps *PositionedStructure) error
+
+// Write creates the filesystem inside the provided image file and populates it
+// with data according to content declartion of the structure. Content data is
+// staged in a temporary location. An optional post-stage helper function can be
+// used to manipulate the data before it is copied over to the image.
+func (f *FilesystemImageWriter) Write(fname string, postStage PostStageFunc) error {
+	st, err := os.Stat(fname)
+	if err != nil {
+		return fmt.Errorf("cannot stat image file: %v", err)
+	}
+	if sz := st.Size(); sz != int64(f.ps.Size) {
+		return fmt.Errorf("size of image file %v is different from declared structure size %v", sz, f.ps.Size)
+	}
+
+	mkfsWithContent := mkfsHandlers[f.ps.Filesystem]
+	if mkfsWithContent == nil {
+		return fmt.Errorf("internal error: filesystem %q has no handler", f.ps.Filesystem)
+	}
+
+	stagingDir := filepath.Join(f.workDir, fmt.Sprintf("snap-stage-content-part-%04d", f.ps.Index))
+	if osutil.IsDirectory(stagingDir) {
+		return fmt.Errorf("cannot prepare staging directory %s: path exists", stagingDir)
+	}
+
+	if err := os.Mkdir(stagingDir, 0755); err != nil {
+		return fmt.Errorf("cannot prepare staging directory: %v", err)
+	}
+
+	if os.Getenv("SNAP_DEBUG_IMAGE_NO_CLEANUP") == "" {
+		defer func() {
+			if err := os.RemoveAll(stagingDir); err != nil {
+				logger.Noticef("cannot remove filesystem staging directory %q: %v", stagingDir, err)
+			}
+		}()
+	}
+
+	mrw, err := NewMountedFilesystemWriter(f.contentDir, f.ps)
+	if err != nil {
+		return fmt.Errorf("cannot prepare filesystem writer for %v: %v", f.ps, err)
+	}
+	// drop all contents to the staging directory
+	if err := mrw.Write(stagingDir, nil); err != nil {
+		return fmt.Errorf("cannot prepare filesystem content: %v", err)
+	}
+
+	if postStage != nil {
+		if err := postStage(stagingDir, f.ps); err != nil {
+			return fmt.Errorf("post stage callback failed: %v", err)
+		}
+	}
+
+	// create a filesystem with contents of the staging directory
+	if err := mkfsWithContent(fname, f.ps.Label, stagingDir); err != nil {
+		return fmt.Errorf("cannot create %q filesystem: %v", f.ps.Filesystem, err)
+	}
+
+	return nil
+}
+
+// NewFilesystemImageWriter returns a writer capable of creating filesystem
+// images corresponding to the provided positiioned structure, with content from
+// the given content directory. A staging directory will be created in
+// optionally prided work directory, otherwise the default temporary storage
+// location will be used.
+func NewFilesystemImageWriter(contentDir string, ps *PositionedStructure, workDir string) (*FilesystemImageWriter, error) {
```

Comment:
(this is probably the worst nitpick ever, sorry!) I like it when the "constructor" is close to the struct definition. Mostly because when reading top to bottom its a bit of a more "natural" flow, i.e. I see what is initialized, validated and then I see how the other methods (like Write) operate on this data. But again, a nickpick^99.

Reply:
Moved